### PR TITLE
Rework of #110 and #69

### LINKED ISSUE #69: Allow selecting the nth matching form
Functions like
```
   $mech->form_with_fields( @fields )
       Selects a form by passing in a list of field names it must contain.  If
       there is more than one form on the page with that matches, then the
       first one is used, and a warning is generated.
```
just "assume" we only want the first matching form. As far as the 2nd
etc. matches, well it will just generate a warning about that.

It really needs an additional parameter, nth, in case the user wants
a match different that the first. For which alas he is currently out
of luck.

### PR DESCRIPTION
**Pull request by simbabque**
I've taken the work done in #110, added the interface changes suggested by @skaji and amended the tests.

To pick up on the original discussion as to what makes up an HTML `name` attribute:

- `name` can be [CDATA](https://www.w3.org/TR/html401/types.html#type-cdata), which means single whitespaces and any characters. Numbers as names are allowed. See https://www.w3.org/TR/html401/interact/forms.html#adef-name-INPUT 
- `id` can contain any character but whitespace, see https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id

Closes #110 
Fixes #69

### CODE REVIEW

**Review by oalders — APPROVED** [review]
🌟

**Comment by simbabque** [comment]
@skaji if you have a minute, I'd appreciate if you could give this a quick glance, as you suggested the design originally. Thanks!

File: lib/WWW/Mechanize.pm [hunk 1]
```diff
@@ -1722,7 +1756,15 @@ sub form_with_fields {
...
     die 'no fields provided' unless scalar @fields;
 
+    my $nth;
+    if ( @fields > 1 && ref $fields[-1] eq 'HASH' ) {
+        $nth = ( pop @fields )->{n};
+    }
+
     my @matches = $self->all_forms_with_fields(@fields);
+    if ( $nth ) {
+      @matches = ( @matches >= $nth ) ? splice(@matches, $nth-1, 1) : ();
```
Comment (skaji) [inline lib/WWW/Mechanize.pm:1766]: nitpick I think
```
($matches[$nth-1])
```
is more straightforward than
```
splice(@matches, $nth-1, 1)
```

File: lib/WWW/Mechanize.pm [hunk 2]
```diff
@@ -1770,21 +1813,45 @@ instead.)
...
+    my $nth;
+    if ( @args % 2 && ref $args[-1] eq 'HASH' ) {
+        $nth = ( pop @args )->{n};
+    }
+
+    my %spec = @args;
+
     my @forms = $self->all_forms_with(%spec);
+    if ( $nth ) {
+      @forms = ( @forms >= $nth ) ? splice(@forms, $nth-1, 1) : ();
```
Comment (skaji) [inline lib/WWW/Mechanize.pm:1853]: nitpick: I think
```
($forms[$nth-1])
```
is more straightforward than

```
splice(@forms, $nth-1, 1)
```

**Review by skaji — APPROVED** [review]
LGTM, thanks!

File: lib/WWW/Mechanize.pm (hunk 2 above)
Comment (simbabque) [reply lib/WWW/Mechanize.pm:1853]: Good shout, thanks. I tried to retain as much as possible from the original commits, but this is much more readable.

**Review by oalders — APPROVED** [review]
👍🏻

**Pull request merged**